ultravox/tools/ds_tool/ds_tool.py

```diff
@@ -223,6 +228,124 @@ def __post_init__(self):
+            ds_chunk_name = f"chunk-range-{chunk_start:09d}-{chunk_end:09d}.parquet"
+            ds_chunk_hub_path = os.path.join(
+                str(self.args.upload_subset), split_name, ds_chunk_name
+            )
+            ds_chunk_cache_path = os.path.join(
+                self.cache_dir,
+                self.args.dataset_name.replace("/", "__"),
+                str(self.args.upload_subset),
+                split_name,
+                ds_chunk_name,
+            )
+            try:
+                if os.path.exists(ds_chunk_cache_path):
+                    print(
+                        f"Skipping chunk {ds_chunk_name} as it has already been processed and uploaded."
+                    )
+                    ds_chunk_processed = datasets.Dataset.from_parquet(
+                        ds_chunk_cache_path
+                    )
+                else:
+                    print(f"Processing chunk {ds_chunk_name}")
+                    ds_chunk_processed = self._process(ds_chunk)
+                    self._upload(ds_chunk_processed, ds_chunk_hub_path, split_name)
+                    ds_chunk_processed.to_parquet(ds_chunk_cache_path)
+                self.total_samples_processed[split_name] = (
+                    self.total_samples_processed.get(split_name, 0)
+                    + len(ds_chunk_processed)
+                )
+
+            except Exception as e:
+                print(f"Failed to upload chunk {ds_chunk_name}: {e}. Retrying later.")
+                if total_chunks == 1:
+                    print(
+                        f"Finished processing and uploading 0/1 chunks for range [{start_index}, {end_index})"
+                    )
+                    self.chunks_not_uploaded.append((start_index, end_index))
+                    return []
+                failed_chunk_ranges.append((chunk_start, chunk_end))
+        successful_chunks = self.args.max_chunk_split - len(failed_chunk_ranges)
+        print(
+            f"Finished processing and uploading {successful_chunks}/{self.args.max_chunk_split} chunks for range [{start_index}, {end_index})"
+        )
+        return failed_chunk_ranges
+
+    def _process(self, ds_chunk: datasets.Dataset) -> datasets.Dataset:
+        return self.args.task.map_split(
+            ds_chunk,
+            self.args.num_workers,
+            self.args.writer_batch_size,
+            self.args.exclude_fields,
+        )
+
+    def _upload(self, ds_chunk_processed: datasets.Dataset, data_dir: str, split_name):
+        print(f"Uploading chunk to hub: {data_dir}")
+        hub_args: Dict[str, Any] = {
+            "config_name": self.args.upload_subset,
+            "token": self.args.token or os.environ.get("HF_TOKEN"),
+            "private": self.args.private,
+            "data_dir": data_dir,
+            "num_shards": self.args.num_shards,
+            "split": split_name,
+        }
+        ds_chunk_processed.push_to_hub(self.args.upload_name, **hub_args)
```

Comment:
Are you absolutely sure this doesn't override the split? I'm fairly confident that it would. Even if it was "appending" to it, `num_shards` would be totally wrong I think.

Reply:
So because I'm uploading it via data_dir, the only thing its overriding is the README (which I should leave a note that we need to update at the end of this whole thing to reflect the correct paths and split metadata).

Reply:
num_shards would be correct for each chunk of data. but total num shards would be num_chunks*num_shards

Reply:
Oh if you have a way of updating README then it's fine.
The num_shards thing is a bit confusing tbh, but that's fine too. Just make sure to clearly comment this where the parameter is defined.

Reply:
Yeah, I will add in a comment to update the README